### PR TITLE
In params().combine() types, flatten unions of interfaces before merging them

### PR DESCRIPTION
**Pull request by kainino0x**
Bug: https://github.com/gpuweb/cts/pull/238#issuecomment-659086834

Changes the type of
```ts
const p = params()
  .combine([{ a: 1 }])
  .combine([
    ...poptions('b', [1]), //
    ...poptions('c', [1]),
  ])
```
from `ParamsBuilder<{ a: number }>`
to `ParamsBuilder<{ a: number, b: number | undefined, c: number | undefined }>`.

### CODE REVIEW

**Comment by kainino0x** [comment]
Opening tentatively, might want to figure out whether this is actually correct before merging though

**Comment by kainino0x** [comment]
@austinEng I can't quite get this working the way I want, but PTAL

**Comment by austinEng** [comment]
err whoops I just saw this

**Comment by kainino0x** [comment]
No problem, it changed completely

**Comment by austinEng** [comment]
to make sure I'm understanding, this is a different goal than:
```typescript
const p = params()
  .combine([{ a: 1 }])
  .combine(poptions('b', [1, 2]))
  .combine(poptions('c', [3, 4]))
  ])
```

which would yield `{ a: 1, b: 1, c: 3}, { a: 1, b: 1, c: 4}, { a: 1, b: 2, c: 3}, { a: 1, b: 2, c: 4}`

```typescript
const p = params()
  .combine([{ a: 1 }])
  .combine([...poptions('b', [1, 2]), ...poptions('c', [3, 4])])
  ])
```

looks like `{ a: 1, b: 1}, { a: 1, b: 2}, { a: 1, c: 3}, { a: 1, c: 4}`

File: src/common/framework/params_builder.ts [hunk 1]
```diff
@@ -1,15 +1,98 @@
...
+ *
+ * And also works for `{ a: number } | { b: string }` which maps to the same.
+ */
+type FlattenUnionOfInterfaces<T> = {
+  [K in AllKeysFromUnion<T>]: KeyOfOr<
+    T,
+    // If T always has K, just take T[K] (union of C[K] for each component C of T):
+    K,
+    // Otherwise, take the union of C[K] for each component C of T, PLUS undefined:
+    undefined | KeyOfOr<UnionToIntersection<T>, K, void>
```
Comment (austinEng) [inline src/common/framework/params_builder.ts:47]: Is the following correct?
 - K is only a keyof the union T if it is present in all of the union's subtypes,
 - if we're in the default case, it means K was absent in at least one of T's subtypes
 - so UnionToIntersection<T>[K] will be never?

I believe these are the cases you describe in your comment "Unexpected test results - hopefully okay to ignore these"

Comment (austinEng) [reply src/common/framework/params_builder.ts:47]: Okay nvm I see it's only a problem if in UnionToIntersection the types conflict -- that's when you'll get never. but if they're optional (undefined) then it's ok.

Comment (kainino0x) [reply src/common/framework/params_builder.ts:47]: Yup. `keyof ({a: number} | {b: string})` is `never` but `keyof ({a: number} & {b: string})` is `'a' | 'b'` I think. for some reason.

**Comment by kainino0x** [comment]
> to make sure I'm understanding

👍

**Comment by kainino0x** [comment]
Types can recurse, so if there was a way to decompose (i.e. `reduce()` over) union types, I think you could do that instead of variadic. But I don't think you can do a pattern match to decompose a union into `First | ...Rest`.

**Pull request merged**